internal/session/session_connect_with.go

```diff
@@ -31,5 +32,8 @@ func (c ConnectWith) validate() error {
 	if c.EndpointTcpPort == 0 {
 		return errors.NewDeprecated(errors.InvalidParameter, op, "missing endpoint ctp port")
 	}
+	if c.UserClientIp == "" {
```

Comment:
This is one of the places that made me wonder about the consequences of the default of `::`. If there's no IP here, we shouldn't have a session connection right?

My guess is that the default is to handle the upgrade case. But I wonder if a better upgrade case is a non-null empty string as a less surprising possibility?

Reply:
`::` means something contextually (unspecified) and an empty string could be interpreted that the zero value was sent to the database.  I'd prefer to keep the `::` default.

Reply:
Why not `0.0.0.0` though? After all, it's used to specify an invalid or unknown address.

We don't know if it was IPv4 or IPv6 -- and it's probably the former -- and while we could just use that and say "if we see this, we didn't know the address at all, regardless of protocol", it technically means that we didn't know _the IPv6_ address.

That's why I had suggested empty instead, because to me it better conveys that we don't know _any_ address, IPv4 or IPv6.

Reply:
`0.0.0.0` is part of any spec that I'm aware of to represent an unspecified IP address.   Empty string is an anti-pattern for unknown value. With that said, we also aren't big fans of null fields in Boundary for required fields, for very good reasons.   Unfortunately, it this case we're going to need to compromise given that we have existing rows.  @mgaffney suggested `::` as a compromise and I'm in agreement that it's better than allowing nulls and better than using an empty string which can lead to all sorts of odd errors since empty string is a zero value.

Reply:
0.0.0.0 is incidentally specified in the DOCSIS standard (https://datatracker.ietf.org/doc/html/rfc4639) to mean unknown address. But incidental.

On the other hand, explicitly making it a v6 spec value means you need to be careful if ever doing comparisons. How do you decide that's an unknown value in the context of IPv4? You're changing up the lack of a standard in v4 with a standard from a different domain.

My main issue is actually not internally to Boundary -- it's surfacing this to users. If most users are only using IPv4 and they are returned an IPv6 address in place of an unknown address, this will be confusing and unexpected. And my assumption is that they will, because it'll surface when doing a read of a session, and it will surface in the warehouse. If this info goes to the warehouse I think it is very much not expected that I as a user need to formulate queries with an IPv6 address when my entire infra is IPv4.

Reply:
Ty for the background about your concerns.  I think, I understand them now which is very helpful.  

I guess, we have a couple of choices to move forward:
- Allow nulls with no default value
- Don't allow nulls and set a default to an address that represents unspecified like `::`, `0.0.0.0` or an empty string

Based on @jefferai concerns, it feels like `::` and `0.0.0.0` may not be ideal and could confuse users.  

At this point, I'd vote for (in order of preference): `null` or an empty string.

@mgaffney ???

Reply:
I know `null` is meant as "value unspecified". What's the difference from a DB modeling perspective between that and "value unknown"?

At an API level null in the DB vs empty string doesn't really matter. But I think there might be some effects in terms of the warehouse if it's `null`...?

Reply:
I vote we use `0.0.0.0` as the default value and not `null`.

Reply:
Just circling back for history -- we realized it must be able to be null because we don't know the value at the time the session is created, only after the worker has filled it in.